src/components/api-tokens/controllers.tsx

```diff
@@ -0,0 +1,353 @@
+  });
+
+  const organization = await cf.organization(params.organizationGUID);
+
+  await checkIfAuthorised(ctx, cf, organization.metadata.guid);
+
+  const users = await cf.usersForOrganization(organization.metadata.guid);
+  const tokens = users.filter(user => isTokenUser(organization.entity.name, user.entity.username));
+
+  const template = new Template(ctx.viewContext, 'API Tokens');
+  template.breadcrumbs = fromOrg(ctx, organization, [
+    { text: 'API Tokens' },
+  ]);
+
+  return {
+    body: template.render(
+      <ListTokens
+        linkTo={ctx.linkTo}
+        organization={organization}
+        tokens={tokens}
+      />,
+    ),
+  };
+}
+
+export async function confirmRevocation(ctx: IContext, params: IParameters): Promise<IResponse> {
+  const cf = new CloudFoundryClient({
+    accessToken: ctx.token.accessToken,
+    apiEndpoint: ctx.app.cloudFoundryAPI,
+    logger: ctx.app.logger,
+  });
+
+  const uaa = new UAAClient({
+    apiEndpoint: ctx.app.uaaAPI,
+    clientCredentials: {
+      clientID: ctx.app.oauthClientID,
+      clientSecret: ctx.app.oauthClientSecret,
+    },
+  });
+
+  const organization = await cf.organization(params.organizationGUID);
+
+  await checkIfAuthorised(ctx, cf, organization.metadata.guid);
+
+  const token = await uaa.getUser(params.tokenGUID);
+  if (!token || !isTokenUser(organization.entity.name, token.userName)) {
+    throw new NotFoundError('token not found');
+  }
+
+  const template = new Template(ctx.viewContext, 'Are you sure you\'d like to revoke the following API Token?');
+  template.breadcrumbs = fromOrg(ctx, organization, [
+    {
+      href: ctx.linkTo('admin.organizations.tokens.list', { organizationGUID: organization.metadata.guid }),
+      text: 'API Tokens',
+    },
+    { text: tokenName(organization.entity.name, token.userName) },
+  ]);
+
+  return {
+    body: template.render(
+      <ConfirmAction
+        action="revoke"
+        csrf={ctx.viewContext.csrf}
+        linkTo={ctx.linkTo}
+        organization={organization}
+        token={token}
+      />,
+    ),
+  };
+}
+
+export async function revoke(ctx: IContext, params: IParameters): Promise<IResponse> {
+  const cf = new CloudFoundryClient({
+    accessToken: ctx.token.accessToken,
+    apiEndpoint: ctx.app.cloudFoundryAPI,
+    logger: ctx.app.logger,
+  });
+
+  const uaa = new UAAClient({
+    apiEndpoint: ctx.app.uaaAPI,
+    clientCredentials: {
+      clientID: ctx.app.oauthClientID,
+      clientSecret: ctx.app.oauthClientSecret,
+    },
+  });
+
+  const organization = await cf.organization(params.organizationGUID);
+
+  await checkIfAuthorised(ctx, cf, organization.metadata.guid);
+
+  const token = await uaa.getUser(params.tokenGUID);
+  if (!token || !isTokenUser(organization.entity.name, token.userName)) {
+    throw new NotFoundError('token not found');
+  }
+
+  await cf.setOrganizationRole(organization.metadata.guid, token.id, 'users', false);
+  await uaa.deleteUser(token.id);
```

Comment:
My preference is to remove them from the organisation but not delete the user, if we can avoid giving Pazmin `uaa.admin`.

We might just need `scim.write` if we locked the user instead. Although I think removing from the organisation should be sufficient

Reply:
The need for `uaa.admin` is not to delete the user, but to "force password change". I know we discussed it, but I think without that functionality the feature may be problematic to users.

As I've described in the PR comment:

> This however could be problematic when one would want to
replicate the permissions of the deleted user - especially when revoking
the token and not noting what access it used to have.

Essentially, I'm worried this would be more appealing feature and I'm not certain if there is much sense delivering these `API Tokens` when we rotate via `Delete user` followed by `Create user` followed by `Set permissions`.

---

But to address your comment, yes, we can definitely replace `deleteUser` with `detachUser` and get similar behaviour.

Reply:
Password changing is going to be a pretty essential feature. I'd also like to see something in place for handling dead accounts, lest there be a large amount of cruft users that exist in UAA never to be resurrected.

TBH I'm less concerned about giving Pazmin uaa.admin as we are going to be playing an easy on easy off trial accounts epic in the next quarter where if you have the appropriate email domain you can set up a trial account without the involvement of support, 90 days later that org and user are removed automagically.